src/api/controllers/db-controller.ts

```diff
@@ -828,43 +944,173 @@ export async function getTxFromDataStore(
-      );
+  }
+  parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
+  if (dbTransaction.contract_call_function_args) {
+    parsedTx.contract_call.function_args = readClarityValueArray(
+      dbTransaction.contract_call_function_args
+    ).map((c, fnArgIndex) => {
+      const functionArgAbi = functionAbi.args[fnArgIndex++];
+      return {
+        hex: bufferToHexPrefixString(serializeCV(c)),
+        repr: cvToString(c),
+        name: functionArgAbi.name,
+        type: getTypeString(functionArgAbi.type),
+      };
+    });
+  }
+  return parsedTx;
+}
+
+export async function searchTxs(
+  db: DataStore,
+  args: GetTxsArgs | GetTxsWithEventsArgs
+): Promise<TransactionList> {
+  const minedTxs = await getTxsFromDataStore(db, args);
+
+  // filtering out tx_ids that were not mined / found
+  const notMinedTransactions: string[] = args.txIds.filter(
+    txId => !minedTxs.find(minedTx => txId === minedTx.tx_id)
+  );
+
+  const foundOrNotMined: TransactionFound[] = [];
+  const nonCanonical: Transaction[] = [];
+  const mempoolTxs: string[] = [];
+  minedTxs.forEach(tx => {
+    // filtering out mined transactions in canonical chain
+    if (tx.canonical && tx.microblock_canonical) {
+      foundOrNotMined.push({ found: true, result: tx });
     }
-    parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
-    if (dbTx.contract_call_function_args) {
-      parsedTx.contract_call.function_args = readClarityValueArray(
-        dbTx.contract_call_function_args
-      ).map((c, fnArgIndex) => {
-        const functionArgAbi = functionAbi.args[fnArgIndex++];
-        return {
-          hex: bufferToHexPrefixString(serializeCV(c)),
-          repr: cvToString(c),
-          name: functionArgAbi.name,
-          type: getTypeString(functionArgAbi.type),
-        };
-      });
+    // filtering out non canonical
+    if (!tx.canonical && !tx.microblock_canonical) {
+      nonCanonical.push(tx);
+      mempoolTxs.push(tx.tx_id);
     }
-  }
+  });
+
+  // finding transactions that are not mined and are not canonical in mempool.
+  mempoolTxs.push(...notMinedTransactions);
+  const mempoolTxsQuery = await getMempoolTxsFromDataStore(db, {
+    txIds: mempoolTxs,
+    includeUnanchored: args.includeUnanchored,
+  });
+
+  const foundOrNotMempool: TransactionFound[] = mempoolTxsQuery.map(
+    (mtx: Transaction | MempoolTransaction) => {
+      return { found: true, result: mtx } as TransactionFound;
+    }
+  );
+  // such transactions that are not in mempoolTxQuery but do exist in notInCanonical
+  const foundOrNotNonCanonical: TransactionFound[] = nonCanonical
+    .filter(
+      tx =>
+        mempoolTxsQuery.findIndex(
+          (mtx: Transaction | MempoolTransaction) => mtx.tx_id === tx.tx_id
+        ) < 0
+    )
+    .map(tx => {
+      return { found: true, result: tx };
+    });
```

Comment:
Perhaps I don't fully understand the purpose of these loops yet. Why do you need to separate mempool txs from non canonical, etc.? Is that information relevant to the response? It seems like in a few lines below you merge all these arrays together.